### ISSUE TITLE
Object.defineProperty doesn't define properties on 'this'

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#28694]
```js
// @ts-check

class C {
  constructor() {
    Object.defineProperty(this, "x", { value: "hello" });

    this.x.toLowerCase();
  }
}
```

**Expected**: No error.
**Actual**: `Property 'x' does not exist on type 'C'.`

**Comment by xprommer** [comment]
Unfortunately it doesn't work in `TS` as well,
if you run this in `TS` playground (I guess it has the latest `TS` version)
```typescript
let obj = {};
Object.defineProperty(obj, "x", { value: "hello", writable: false });

obj.x.toLowercase();
```
then it will show only one error message:
`"Property 'x' does not exist on type '{}'"`

but I expected to get different error like `"Property 'toLowercase' does not exist on type 'string'"` as was described in https://devblogs.microsoft.com/typescript/announcing-typescript-3-2/#objectdefineproperty-declarations-in-javascript

**Comment by tokland** [comment]
@DanielRosenwasser  See https://github.com/Microsoft/TypeScript/issues/14650

**Comment by DanielRosenwasser** [comment]
@tokland This is about JavaScript analysis, not TypeScript.

**Comment by boneskull** [comment]
I'm running in to this one.  Anyone know of a workaround?

**Comment by boneskull** [comment]
n/m, workaround for above is something like this:

```js
// @ts-check

class C {
  constructor() {
    /** @type {Readonly<string>} */
    this.x = undefined;
    Object.defineProperty(this, "x", { value: "hello" });

    this.x.toLowerCase();
  }
}
```

**Comment by akornatskyy** [comment]
You need to add definite assertion assignment to property 'x: string'.
```ts
class C {
  x!: string;  // definite assertion
  constructor() {
    Object.defineProperty(this, 'x', {value: 'hello'});
    this.x.toLowerCase();
  }
}
```

**Comment by brainkim** [comment]
@akornatskyy I think the source is assumed to be JavaScript in this issue and there is no definite assignment assertion syntax there as far as I can tell.

**Comment by BridgeAR** [comment]
I am running into this issue in a big code base as well. Could this be looked into again / is that something that could be added to a next roadmap @DanielRosenwasser?
